### PR TITLE
Semantic highlighting for unsafe union field access

### PR DESCRIPTION
**Pull request by Nashenas88**
This change adds support for unions in inference and lowering, then extends on that to add the unsafe semantic modifier on field access only. The `is_possibly_unsafe` function in `syntax_highlighting.rs` could be extended to support fns and static muts so that their definitions are not highlighted as unsafe, but only their usage.

Also, each commit of this PR updates the tests. By reviewing the files by commit, it's easy to see how the changes in the code affected the tests.

### CODE REVIEW

File: crates/ra_ide/src/syntax_highlighting.rs [hunk 1]
```diff
@@ -498,8 +519,8 @@ fn highlight_element(
         }
         NAME_REF => {
             let name_ref = element.into_node().and_then(ast::NameRef::cast).unwrap();
-            match classify_name_ref(sema, &name_ref) {
-                Some(name_kind) => match name_kind {
+            match (classify_name_ref(sema, &name_ref), is_possibly_unsafe(&name_ref)) {
```
Comment (jonas-schievink) [inline crates/ra_ide/src/syntax_highlighting.rs:522]: This match doesn't seem to inspect the `is_possibly_unsafe(&name_ref)`, so maybe just put it in a variable before the match?

File: crates/ra_hir_ty/src/infer.rs [hunk 2]
```diff
@@ -440,6 +440,12 @@ impl<'a> InferenceContext<'a> {
                 let ty = self.insert_type_vars(ty.subst(&substs));
                 forbid_unresolved_segments((ty, Some(strukt.into())), unresolved)
             }
+            TypeNs::AdtId(AdtId::UnionId(u)) => {
+                let substs = Ty::substs_from_path(&ctx, path, u.into(), true);
+                let ty = self.db.ty(u.into());
+                let ty = self.insert_type_vars(ty.subst(&substs));
+                forbid_unresolved_segments((ty, Some(u.into())), unresolved)
+            }
```
Comment (jonas-schievink) [inline crates/ra_hir_ty/src/infer.rs:448]: Can you also add an inference test for these changes?

File: crates/ra_ide/src/syntax_highlighting.rs (hunk 1 above)
Comment (Nashenas88) [reply crates/ra_ide/src/syntax_highlighting.rs]: Done

File: crates/ra_hir_ty/src/infer.rs (hunk 2 above)
Comment (Nashenas88) [reply crates/ra_hir_ty/src/infer.rs:448]: I expanded the existing test. I reworked the git history so the test is added before the above change. That helps show that the inference for the union construction was incorrect before this change.

**Pull request merged**